### PR TITLE
[CK-Tile] Universal gemm memory bound pipeline

### PR DESCRIPTION
**Pull request by aosewski**

### CODE REVIEW

**Comment by aosewski** [comment]
@carlushuang, @ThomasNing  Could you please review ? Thanks!

File: example/ck_tile/03_gemm/CMakeLists.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
-set(CMAKE_BUILD_TYPE Debug)
-add_executable(tile_example_gemm_basic EXCLUDE_FROM_ALL gemm_basic.cpp)
\ No newline at end of file
+# set(CMAKE_BUILD_TYPE Debug)
```
Comment (bartekxk) [inline example/ck_tile/03_gemm/CMakeLists.txt]: Do we need this?

File: example/ck_tile/03_gemm/gemm_basic.cpp [hunk 2]
```diff
@@ -280,75 +278,32 @@ int main(int argc, char* argv[])
...
-            {
-                stride_c = N;
-            }
-        }
-
-        ck_tile::HostTensor<CDataType> c_host_gpu_ref(c_dimensions);
+        ck_tile::HostTensor<CDataType> c_host_gpu_ref(
+            f_host_tensor_descriptor(M, N, stride_C, CLayout{}));
         ck_tile::DeviceMem c_gpu_buf(c_host_gpu_ref.get_element_space_size_in_bytes());
+        c_gpu_buf.SetZero();
```
Comment (bartekxk) [inline example/ck_tile/03_gemm/gemm_basic.cpp]: Maybe if (splitK !=1)?

File: example/ck_tile/03_gemm/gemm_basic.hpp [hunk 3]
```diff
@@ -20,7 +18,7 @@ struct GemmBasicTypeConfig<ck_tile::half_t>
     using ADataType   = ck_tile::half_t;
     using BDataType   = ck_tile::half_t;
     using AccDataType = float;
-    using CDataType   = ck_tile::half_t; // type convert
+    using CDataType   = float;
```
Comment (bartekxk) [inline example/ck_tile/03_gemm/gemm_basic.hpp]: Why?

File: example/ck_tile/03_gemm/gemm_basic.cpp [hunk 4]
```diff
@@ -11,6 +10,11 @@
 #include <string>
 #include <tuple>
 
+#include "ck_tile/ops/epilogue.hpp"
```
Comment (bartekxk) [inline example/ck_tile/03_gemm/gemm_basic.cpp:12]: Why you dont include gemm_basic.hpp?

File: example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp [hunk 5]
```diff
@@ -0,0 +1,379 @@
+
```
Comment (bartekxk) [inline example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp]: Pls remove first empty line

File: example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+#include <ostream>
+#include <string>
+#include <tuple>
+
+#include "ck_tile/ops/epilogue.hpp"
+#include "ck_tile/ops/gemm.hpp"
+#include "ck_tile/host.hpp"
+#include "gemm_basic.hpp"
+
+auto create_args(int argc, char* argv[])
```
Comment (bartekxk) [inline example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp]: Could you  move this to gemm_basic.hpp?

File: example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+    int n_warmup                = arg_parser.get_int("warmup");
+    int n_repeat                = arg_parser.get_int("repeat");
+
+    using ALayout = ck_tile::tensor_layout::gemm::RowMajor;
+    using BLayout = ck_tile::tensor_layout::gemm::ColumnMajor;
+    using CLayout = ck_tile::tensor_layout::gemm::RowMajor;
+
+    using namespace ck_tile::literals;
+
+    auto f_host_tensor_descriptor =
```
Comment (bartekxk) [inline example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp]: I think that large part of this file and gemm_basic.cpp could be moved to gemm_basic.hpp

File: include/ck_tile/host/reference/reference_gemm.hpp [hunk 8]
```diff
@@ -1,61 +1,50 @@
...
-                         ck_tile::type_convert<AccDataType>(v_b);
-            }
+        for(std::size_t k = 0; k < K; ++k)
+        {
+            ADataType v_a = a_element_op(a_m_k(m, k));
+            BDataType v_b = b_element_op(b_k_n(k, n));
 
-            c_m_n(m, n) = ck_tile::type_convert<CDataType>(acc_element_op(v_acc));
+            v_acc +=
+                ck_tile::type_convert<AccDataType>(v_a) * ck_tile::type_convert<AccDataType>(v_b);
```
Comment (bartekxk) [inline include/ck_tile/host/reference/reference_gemm.hpp:41]: It should be casted after compute.

File: include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr.hpp [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+#include "ck_tile/core.hpp"
+#include "ck_tile/ops/gemm/block/block_gemm_as_bs_cr_default_policy.hpp"
+
+namespace ck_tile {
+
+// A is block window on shared memory
+// B is block window on shared memory
+// C is block distributed tensor
+template <typename Problem_, typename Policy_ = BlockGemmAsBsCrDefaultPolicy>
+struct BlockGemmAsBsCr
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr.hpp]: Why AsBs?

File: include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr.hpp [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    using Policy         = remove_cvref_t<Policy_>;
+    using ADataType      = remove_cvref_t<typename Problem::ADataType>;
+    using BDataType      = remove_cvref_t<typename Problem::BDataType>;
+    using CDataType      = remove_cvref_t<typename Problem::CDataType>;
+    using BlockGemmShape = remove_cvref_t<typename Problem::BlockGemmShape>;
+
+    static constexpr index_t kBlockSize = Problem::kBlockSize;
+
+    // C += A * B
+    template <typename CBlockTensor, typename ABlockWindowTmp, typename BBlockWindowTmp>
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr.hpp]: Maybe use _ instead of Tmp

File: include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr.hpp [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+        auto c_block_tensor = make_static_distributed_tensor<CDataType>(c_block_dstr);
+        return c_block_tensor;
+    }
+
+    using CBlockTile = remove_cvref_t<decltype(MakeCBlockTile())>;
+
+    // C = A * B
+    template <typename ABlockTensorTmp, typename BBlockWindowTmp>
+    CK_TILE_DEVICE auto operator()(const ABlockTensorTmp& a_block_tensor_tmp,
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr.hpp]: Could we replace these "tmp"?

File: include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr_default_policy.hpp [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+                         kNPerBlock % 128 == 0 % kKPerBlock % 16 == 0)
+            {
+                return make_tuple(WarpGemmMfmaF16F16F32M32N32K16{}, 2, 2);
+            }
+            else
+            {
+                return make_tuple(WarpGemmMfmaF16F16F32M32N32K16{}, 2, 2);
+            }
+#else
+            return make_tuple(WarpGemmMfmaF16F16F32M32N32K16TransposedCDistribution{}, 4, 1);
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr_default_policy.hpp]: Is it good idea? Can we create warpGemm base BlockWarp?

File: include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp [hunk 13]
```diff
@@ -3,38 +3,34 @@
...
+    using CLayout                            = remove_cvref_t<typename GemmPipeline::CLayout>;
+    static constexpr index_t KernelBlockSize = GemmPipeline::BlockSize;
 
-    using ADataType    = remove_cvref_t<typename GemmPipeline::ADataType>;
-    using BDataType    = remove_cvref_t<typename GemmPipeline::BDataType>;
-    using CAccDataType = remove_cvref_t<typename GemmPipeline::CDataType>;
-    using CODataType   = remove_cvref_t<typename EpiloguePipeline::ODataType>;
+    using ADataType = remove_cvref_t<typename GemmPipeline::ADataType>;
+    using BDataType = remove_cvref_t<typename GemmPipeline::BDataType>;
+    // using CAccDataType = remove_cvref_t<typename GemmPipeline::CDataType>;
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:28]: Why did you remove AccDataType?

File: include/ck_tile/ops/gemm/pipeline/block_gemm_pipeline_problem.hpp [hunk 14]
```diff
@@ -33,4 +40,43 @@ struct BlockGemmPipelineProblem
...
+          typename ALayout_,
+          typename BLayout_,
+          typename CLayout_,
+          bool kPadA_                      = false,
+          bool kPadB_                      = false,
+          bool kPadC_                      = false,
+          GemmPipelineScheduler Scheduler_ = GemmPipelineScheduler::Intrawave,
+          bool HasHotLoop_                 = true,
+          TailNumber TailNum_              = TailNumber::Full>
+struct UniversalGemmPipelineProblem
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/pipeline/block_gemm_pipeline_problem.hpp]: Do we need separate Problem structure for universal gemm? Maybe we can update previous version with some default parameters

File: include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_mem.hpp [hunk 15]
```diff
@@ -0,0 +1,417 @@
...
+    static constexpr index_t BlockSize = Problem::kBlockSize;
+    static constexpr index_t MPerBlock = BlockGemmShape::kM;
+    static constexpr index_t NPerBlock = BlockGemmShape::kN;
+    static constexpr index_t KPerBlock = BlockGemmShape::kK;
+
+    static constexpr index_t WgpPerCU =
+        (4 * get_warp_size() / BlockSize) >= 1 ? 4 * get_warp_size() / BlockSize : 1;
+    // TODO: Is this 32K value gfx9 arch specific?
+    static constexpr index_t FullMemBandPrefetchStages = integer_divide_ceil(
+        32768 / WgpPerCU,
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_mem.hpp]: Can we assign this 32768 to some constexpr variable?

File: include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_mem_custom_policy.hpp [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+namespace ck_tile {
+
+// Default policy for GemmPipelineAGmemBGmemCRegV1
+
+// Maximum Global Memory throughput pipeline with >=32KB data in fly
+// GlobalPrefetchStages: >=2
+// LocalPreFillStages: 1
+// LocalPreFetchStages: 0
+// LocalSharedMemoryBuffer: 1
+struct GemmPipelineAgBgCrMemCustomPolicy
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_mem_custom_policy.hpp]: Why "custom" I dont see any possiblity to customize this policy.

File: include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_mem_custom_policy.hpp [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+
+#include "ck_tile/core.hpp"
+#include "ck_tile/ops/gemm/block/block_gemm_as_bs_cr_default_policy.hpp"
+#include "ck_tile/ops/gemm/block/block_gemm_as_bs_cr.hpp"
+
+namespace ck_tile {
+
+// Default policy for GemmPipelineAGmemBGmemCRegV1
+
+// Maximum Global Memory throughput pipeline with >=32KB data in fly
```
Comment (bartekxk) [inline include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_mem_custom_policy.hpp]: What is the difference of this policy compared to include/ck_tile/ops/gemm/pipeline/block_gemm_pipeline_agmem_bgmem_creg_v1_default_policy.hpp ? Can we customize mentioned block_gemm_pipeline_agmem_bgmem_creg_v1_default_policy  with listed parameters? Maybe we can avoid code duplication

File: test/ck_tile/gemm/test_gemm_mem_pipeline.cpp [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+using KernelTypes = ::testing::Types<
+    //         ALayout, BLayout, CLayout, ADataType, BDataType, AccDataType, CDataType
+    std::tuple<    Row,     Col,     Row,       F16,       F16,         F32,      F32>
+    // TODO: fixme!
+    // std::tuple<    Col,     Row,     Row,       F16,       F16,         F32,      F32>,
+    // std::tuple<    Row,     Row,     Row,       F16,       F16,         F32,      F32>,
+    // std::tuple<    Col,     Col,     Row,       F16,       F16,         F32,      F32>
+    >;
+// clang-format on
+
```
Comment (bartekxk) [inline test/ck_tile/gemm/test_gemm_mem_pipeline.cpp]: why disabled?

**Review by bartekxk — COMMENTED** [review]
Good job!

File: example/ck_tile/03_gemm/gemm_basic.cpp (hunk 2 above)
Comment (aosewski) [reply example/ck_tile/03_gemm/gemm_basic.cpp]: Hmm maybe this won't be necessary all the time, but that's just a precautionary measure.

File: example/ck_tile/03_gemm/gemm_basic.hpp (hunk 3 above)
Comment (aosewski) [reply example/ck_tile/03_gemm/gemm_basic.hpp]: I had some errors with respect to selected MFMA instruction... I'll recheck it.

Comment (aosewski) [reply example/ck_tile/03_gemm/gemm_basic.hpp]: Fixed. Now using AccDataType as MFMA output data type.

File: example/ck_tile/03_gemm/gemm_basic.cpp (hunk 4 above)
Comment (aosewski) [reply example/ck_tile/03_gemm/gemm_basic.cpp:12]: I'm including it at line number 16.

File: include/ck_tile/host/reference/reference_gemm.hpp (hunk 8 above)
Comment (aosewski) [reply include/ck_tile/host/reference/reference_gemm.hpp:41]: No, that's not true. The intermediate accumulation values should be in AccDataType. The output of MFMA is in AccDataType.

File: include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr.hpp (hunk 9 above)
Comment (aosewski) [reply include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr.hpp]: `s` stands for shared memory, that is LDS. `r` stands for registers.

File: include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr_default_policy.hpp (hunk 12 above)
Comment (aosewski) [reply include/ck_tile/ops/gemm/block/block_gemm_as_bs_cr_default_policy.hpp]: I don't quite understand. I've reverted back to use existing gemm policies. I haven't touched the structure of this file in this PR. Will be reconsidered in future PRs.

File: include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp (hunk 13 above)
Comment (aosewski) [reply include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:28]: It's not used anywhere...

File: include/ck_tile/ops/gemm/pipeline/block_gemm_pipeline_problem.hpp (hunk 14 above)
Comment (aosewski) [reply include/ck_tile/ops/gemm/pipeline/block_gemm_pipeline_problem.hpp]: I don't want to mess with FMHA kernels which use the `BlockGemmPipelineProblem`. I'm not considering current state of `UniversalGemmPipelineProblem` structure as final. It may change in next PRs.

File: test/ck_tile/gemm/test_gemm_mem_pipeline.cpp (hunk 18 above)
Comment (aosewski) [reply test/ck_tile/gemm/test_gemm_mem_pipeline.cpp]: They're not working :) Need to fix in future PR

File: include/ck_tile/ops/gemm/pipeline/block_gemm_pipeline_problem.hpp (hunk 14 above)
Comment (ThomasNing) [reply include/ck_tile/ops/gemm/pipeline/block_gemm_pipeline_problem.hpp]: Hi Adam, Cheers for this new PR! Please allow me to take some time to review this PR. I will finish the review by this weekend or next Monday. FYI on this issue. I discussed with Qianfeng and Yao Dan today. I will modify the fmha to let it calling from the Block level problem of the GEMM instead of the Pipeline Problem. I also rename the Pipeline name to delete all the confusing "Block" name in the pipeline file.

**Review by bartekxk — DISMISSED** [review]
LGTM

File: example/ck_tile/03_gemm/gemm_basic.cpp [hunk 19]
```diff
@@ -89,295 +95,20 @@ float gemm_calc(const gemm_basic_args& args, const ck_tile::stream_config& s)
...
-            if constexpr(std::is_same_v<LayoutC, ck_tile::tensor_layout::gemm::ColumnMajor>)
-            {
-                return M;
-            }
-            else
-            {
-                return N;
-            }
-        }();
+        std::cout << "Lunching kernel with args:"
```
Comment (ThomasNing) [inline example/ck_tile/03_gemm/gemm_basic.cpp]: Typo: Launching.

File: example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+#include <string>
+#include <tuple>
+
+#include "ck_tile/ops/epilogue.hpp"
+#include "ck_tile/ops/gemm.hpp"
+#include "ck_tile/host.hpp"
+#include "gemm_basic.hpp"
+
+template <typename ALayout, typename BLayout, typename CLayout>
+float gemm_calc(const gemm_basic_args& args, const ck_tile::stream_config& s)
```
Comment (ThomasNing) [inline example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp]: How you define a basic_mem_pipeline? If we add one more pipeline should I add another example? Basic GEMM could be a sepatate example from other gemm+ operations but all of the basic gemm pipeline methods should have a file so that it is easy to control by the future ck tile engine.

File: example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+                                       args.stride_A,
+                                       args.stride_B,
+                                       args.stride_C);
+
+        const dim3 grids      = Kernel::GridSize(args.M, args.N, args.kbatch);
+        constexpr dim3 blocks = Kernel::BlockSize();
+
+        if(s.log_level_ > 0)
+        {
+            std::cout << "Lunching kernel with args:"
```
Comment (ThomasNing) [inline example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp]: Same "Launching".

File: include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_mem.hpp [hunk 22]
```diff
@@ -0,0 +1,417 @@
...
+    using ADataType      = remove_cvref_t<typename Problem::ADataType>;
+    using BDataType      = remove_cvref_t<typename Problem::BDataType>;
+    using BlockGemmShape = remove_cvref_t<typename Problem::BlockGemmShape>;
+
+    static constexpr index_t BlockSize = Problem::kBlockSize;
+    static constexpr index_t MPerBlock = BlockGemmShape::kM;
+    static constexpr index_t NPerBlock = BlockGemmShape::kN;
+    static constexpr index_t KPerBlock = BlockGemmShape::kK;
+
+    static constexpr index_t MinMemInFlyBytes = 32768;
```
Comment (ThomasNing) [inline include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_mem.hpp:28]: Could we do that in a macro  instead of hard-coding that?

File: test/ck_tile/gemm/test_gemm_mem_pipeline_util.hpp [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+        ck_tile::index_t kbatch;
+        ck_tile::index_t M;
+        ck_tile::index_t N;
+        ck_tile::index_t K;
+        ck_tile::index_t stride_A;
+        ck_tile::index_t stride_B;
+        ck_tile::index_t stride_C;
+    };
+
+    void invoke_gemm(const gemm_basic_args& args, const ck_tile::stream_config& s)
```
Comment (ThomasNing) [inline test/ck_tile/gemm/test_gemm_mem_pipeline_util.hpp:41]: The invoke_gemm function and the other test function is really similar to the example section. Is there a way we could calling the same function?

File: example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp (hunk 20 above)
Comment (aosewski) [reply example/ck_tile/03_gemm/gemm_basic_mem_pipeline.cpp]: This is temporary solution to test new provided functionality.

File: test/ck_tile/gemm/test_gemm_mem_pipeline_util.hpp (hunk 23 above)
Comment (aosewski) [reply test/ck_tile/gemm/test_gemm_mem_pipeline_util.hpp:41]: No. I do not want to add `gtest` dependency to examples. This is the outcome of our current examples design, which contain also host API.

**Review by bartekxk — APPROVED** [review]
### LGTM
